### PR TITLE
feat: add lru caching algorithm

### PR DESCRIPTION
**Pull request by foo290**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1557"></a>

### CODE REVIEW

File: others/lru_cache.cpp [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+ * @namespace lru_cache
+ * @brief Implementation of LRU caching algorithm
+ */
+namespace lru_cache {
+
+/**
+ * @brief LRU cache class
+ */
+class LRUCache {
+    int pageFrame;         ///< Page frame, or total size of the cache.
```
Comment (Panquesito7) [inline others/lru_cache.cpp:63]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). 🙂

File: others/lru_cache.cpp [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+ * */
+
+#include <cassert>        /// for assert
+#include <iostream>       /// for IO Operations
+#include <list>           /// for std::list
+#include <unordered_map>  /// for std::unordered_map
+
+/**
+ * @namespace lru_cache
+ * @brief Implementation of LRU caching algorithm
```
Comment (Panquesito7) [inline others/lru_cache.cpp:55]:
```suggestion
 * @brief Implementation of the [LRU caching algorithm](https://en.wikipedia.org/wiki/Cache_replacement_policies#Least_recently_used_(LRU))
```

File: others/lru_cache.cpp [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+ */
+namespace lru_cache {
+
```
Comment (Panquesito7) [inline others/lru_cache.cpp:58]:
```suggestion
```

File: others/lru_cache.cpp [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+ *
+ * @author [Nitin Sharma](https://github.com/foo290)
+ * */
+
+#include <cassert>        /// for assert
+#include <iostream>       /// for IO Operations
+#include <list>           /// for std::list
+#include <unordered_map>  /// for std::unordered_map
+
+/**
```
Comment (Panquesito7) [inline others/lru_cache.cpp:53]:
```suggestion
/**
 * @namespace others
 * @brief Other algorithms
 */
namespace others {
/**
```

File: others/lru_cache.cpp [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+     * */
+    int getHits() const { return hits; }
+    /**
+     * @brief A function to get page fault
+     * @returns int
+     * */
+    int getPageFault() const { return pageFault; }
+};
+
+}  // namespace lru_cache
```
Comment (Panquesito7) [inline others/lru_cache.cpp:132]:
```suggestion
}  // namespace others
}  // namespace lru_cache
```

File: others/lru_cache.cpp [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+ * miss
+ * @returns void
+ * */
+static void test_1() {
+    int expected_hits = 2;
+    int expected_pageFault = 4;
+
+    log("Running Test-1...");
+
+    lru_cache::LRUCache cache(4);
```
Comment (Panquesito7) [inline others/lru_cache.cpp:158]:
```suggestion
    others::lru_cache::LRUCache cache(4);
```

File: others/lru_cache.cpp [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+ * miss
+ * @returns void
+ * */
+static void test_2() {
+    int expected_hits = 4;
+    int expected_pageFault = 2;
+
+    log("Running Test-2...");
+
+    lru_cache::LRUCache cache(4);
```
Comment (Panquesito7) [inline others/lru_cache.cpp:185]:
```suggestion
    others::lru_cache::LRUCache cache(4);
```

File: others/lru_cache.cpp [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+ * miss
+ * @returns void
+ * */
+static void test_3() {
+    int expected_hits = 1;
+    int expected_pageFault = 5;
+
+    log("Running Test-3...");
+
+    lru_cache::LRUCache cache(4);
```
Comment (Panquesito7) [inline others/lru_cache.cpp:212]:
```suggestion
    others::lru_cache::LRUCache cache(4);
```

File: others/lru_cache.cpp [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+    test_3();
+    log("");
+    log("TESTS COMPLETED!");
+}
+}  // namespace lru_tests
+
+/**
+ * @brief Main function
+ * @param argc commandline argument count (ignored)
+ * @param argv commandline array of arguments (ignored)
```
Comment (Panquesito7) [inline others/lru_cache.cpp:243]:
```suggestion
```

File: others/lru_cache.cpp [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+ * @returns 0 on exit
+ */
+int main() {
+    lru_tests::run_tests();
+
+    // Usage
+    lru_cache::LRUCache cache(4);
```
Comment (Panquesito7) [inline others/lru_cache.cpp:250]:
```suggestion
    others::lru_cache::LRUCache cache(4);
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Good work!

File: others/lru_cache.cpp [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+ *          * If the cache is full, REMOVE the last element from the cache.
+ *          * If the element is present in cache, then shift that element to
+ * first position in cache from its original position.
+ *              * This way you can keep the least recently used elements in the
+ * last and most recently used in front of the cache.
+ *
+ *  Every time a requested page is  not found in cache, that is a miss or page
+ * fault, and if the page is present in cache, then its a hit.
+ *
+ * ## Data Structure used:
```
Comment (Panquesito7) [inline others/lru_cache.cpp:33]:
```suggestion
 * ## Data Structure used
```

File: others/lru_cache.cpp [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+ * @namespace others
+ * @brief Other algorithms
+ */
+namespace others {
+/**
+ * @namespace lru_cache
+ * @brief Implementation of LRU caching algorithm
+ */
+namespace lru_cache {
+
```
Comment (Panquesito7) [inline others/lru_cache.cpp:63]:
```suggestion
```

File: others/lru_cache.cpp [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+#include <unordered_map>  /// for std::unordered_map
+
+/**
+ * @namespace others
+ * @brief Other algorithms
+ */
+namespace others {
+/**
+ * @namespace lru_cache
+ * @brief Implementation of LRU caching algorithm
```
Comment (Panquesito7) [inline others/lru_cache.cpp:60]:
```suggestion
 * @brief Implementation of the [LRU caching algorithm](https://en.wikipedia.org/wiki/Cache_replacement_policies#Least_recently_used_(LRU))
```

File: others/lru_cache.cpp [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+    uint64_t getHits() const { return hits; }
+    /**
+     * @brief A function to get page fault
+     * @returns int
+     * */
+    uint64_t getPageFault() const { return pageFault; }
+};
+
+}  // namespace lru_cache
+}  // namespace others
```
Comment (Panquesito7) [inline others/lru_cache.cpp:140]:
```suggestion
}  // namespace others
}  // namespace lru_cache
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
A tip: don't accept the reviewer's suggestions that quickly, else the `github-actions` bot will force-push the repository and overwrite those changes.

File: others/lru_cache.cpp (hunk 14 above)
Comment (foo290) [reply others/lru_cache.cpp:140]: I think this is right, it's a clang-tidy fix, the first namespace is others and the second is lru_cache, so lru_cache will be closed before others.

**Comment by foo290** [comment]
> A tip: don't accept the reviewer's suggestions that quickly, else the `github-actions` bot will force-push the repository and overwrite those changes.

oh oky, thanks 😃

File: others/lru_cache.cpp (hunk 14 above)
Comment (Panquesito7) [reply others/lru_cache.cpp:139]: Ohh, I got confused when I made that review. Sorry about that. 😅

File: others/lru_cache.cpp [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+    cache.refer(2);
+    cache.refer(3);
+    cache.refer(4);
+    cache.refer(5);
+    cache.refer(5);
+
+    cache.display();
+
+    std::cout << "Hits: " << cache.getHits()
+              << " Miss: " << cache.getPageFault();
```
Comment (Panquesito7) [inline others/lru_cache.cpp:266]:
```suggestion
              << " Miss: " << cache.getPageFault() << std::endl;
```

**Review by Panquesito7 — APPROVED** [review]
Awesome work, @foo290! Thank you for your contribution! 😄👍🎉

**Review by mishraabhinn — APPROVED** [review]

**Pull request merged**